openwpm/storage/cloud_storage/s3_storage.py

```diff
@@ -1,5 +1,103 @@
+import logging
+from typing import Any, Set
+
+import pyarrow.parquet as pq
+from pyarrow.lib import Table
+from s3fs import S3FileSystem
+
 from ..arrow_storage import ArrowProvider
+from ..storage_providers import TableName, UnstructuredStorageProvider
+
+
+class S3StructuredProvider(ArrowProvider):
+    """This class allows you to upload Parquet files to S3.
+
+    Inspired by the old S3Aggregator structure the S3StructuredProvider
+    will by default store into
+    base_path/visits/table_name in the given bucket.
+
+    Pass a different sub_dir to change this.
+
+    **kwargs get passed on to S3FileSystem.__init__
+    Please look at https://s3fs.readthedocs.io/en/latest/api.html#s3fs.core.S3FileSystem
+    for further information
+    """
+
+    file_system: S3FileSystem
+
+    def __init__(
+        self, bucket_name: str, base_path: str, sub_dir: str = "visits", **kwargs: Any
+    ) -> None:
+        super().__init__()
+        self.kwargs = kwargs
+        self.base_path = f"{bucket_name}/{base_path}/{sub_dir}/{{table_name}}"
+
+    def __str__(self) -> str:
+        return f"S3FS:{self.base_path.removesuffix('/{table_name}')}"
+
+    async def init(self) -> None:
+        await super(S3StructuredProvider, self).init()
+        self.file_system = S3FileSystem(**self.kwargs)
+
+    async def write_table(self, table_name: TableName, table: Table) -> None:
+        self.file_system.start_transaction()
+        pq.write_to_dataset(
+            table,
+            self.base_path.format(table_name=table_name),
+            filesystem=self.file_system,
+        )
+        self.file_system.end_transaction()
+
+    async def shutdown(self) -> None:
```

Comment:
Are you sure we want to pass here? This overwrites ArrowProvider's default method which logs an error if there are any unwritten batches. That still seems useful.

I _think_ we can just drop this.